extract.py

```diff
@@ -0,0 +1,408 @@
+        spec_box_0 = spectrum.sum(axis = 0)                            #initial box-extracted spectrum
+        var_box_0 = var.sum(axis = 0)                                #initial variance guess
+
+        newM = np.ones_like(spectrum)                                #not masking any pixels because we interpolated over them
+
+        if convert_to_bool(obs_par['opt_extract'])==True: [f_opt_0, var_opt_0, numoutliers] = optextr.optextr(spectrum, err, spec_box_0, var_box_0, newM, ancil.nsmooth, ancil.sig_cut, ancil.diagnostics)
+        else: [f_opt, var_opt] = [spec_box_0,var_box_0]
+
+        #sums up spectra and variance for all the differenced images
+        spec_opt += f_opt_0
+        var_opt += var_opt_0
+        spec_box += spec_box_0
+        var_box += var_box_0
+
+    ######################################################################################################################################
+
+    time = (d[0].header['expstart'] + d[0].header['expend'])/(2.0) + 2400000.5                    #converts time to BJD_TDB; see Eastman et al. 2010 equation 4
+    time = time + (32.184)/(24.0*60.0*60.0)
+    time = time + (suntimecorr.suntimecorr(ancil.ra, ancil.dec, array([time]), ancil.coordtable[ancil.visnum], verbose=False))/(60.0*60.0*24.0)
+
+    phase = (time-ancil.t0)/ancil.period - math.floor((time-ancil.t0)/ancil.period)
+    if phase > 0.5: phase = phase - 1.0                                    #calculates orbital phase
+
+    template_waves = ancil.wave_grid[0, int(ancil.refpix[0,1]) + ancil.LTV1, cmin:cmax]             #LK interpolation 8/18
+    shift = 0.
+    #corrects for wavelength drift over time
+    if convert_to_bool(obs_par['correct_wave_shift']) == True:
+        if nspectra == 0:
+            shift = 0.
+            template_spectrum = spec_opt                                #makes the first exposure the template spectrum
+            best_spec = spec_opt
+            best_var = var_opt
+        else:
+            #shifts spectrum so it matches the template
+            [best_spec, best_var, shift] = interpolate_spectrum(spec_opt, np.sqrt(var_opt), template_spectrum, template_waves)
+            best_var = best_var**2
+
+            spec_opt = best_spec                                    #saves the interpolated spectrum
+            var_opt = best_var
+
+    #if convert_to_bool(obs_par['plot_spectrum']) == True: plot_spectrum(ancil)
+    #print sum(spec_opt)/sum(var_opt), sum(spec_box)/sum(var_box), sum(spec_opt)/sum(spec_box)
+
+    """plt.plot(template_waves, cRate)
+    plt.plot(template_waves, cRate1)
+    plt.plot(template_waves, cRate2)
+    plt.show()"""
+
+    #plt.plot(template_waves, spec_opt)
+    #plt.show()
+
+    n = len(spec_opt)
+    print(phase[0], sum(spec_opt), sum(var_opt),  sum(spec_box), sum(var_box), time[0], ancil.visnum, ancil.orbnum, scan)
+    if ancil.output == True:
```

Comment:
Please include a header for these files, user needs to find this code to understand the columns.